### PR TITLE
feat(api): impl some web3 interfaces

### PR DESCRIPTION
**Pull request by DavidChild**
**Which docs this PR relation:**

**Which toolchain this PR adaption:**
1. eth_coinbase
2. eth_hashrate
3. eth_submitWork
4. eth_submitHashrate

**No Breaking Change**

**Special notes for your reviewer:**

### CODE REVIEW

**Review by driftluo — CHANGES_REQUESTED** [review]
must not import `async-std`

File: core/api/src/adapter.rs [hunk 1]
```diff
@@ -200,4 +200,8 @@ where
             .get(hash.as_bytes())?
             .ok_or_else(|| APIError::Adapter("Can't find this position".to_string()).into())
     }
+
+    async fn get_number_by_hash(&self, ctx: Context, hash: Hash) -> ProtocolResult<Option<u64>> {
```
Comment (KaoImin) [inline core/api/src/adapter.rs]: Remove this interface.

File: core/api/src/jsonrpc/impl.rs [hunk 2]
```diff
@@ -681,6 +698,7 @@ fn from_receipt_to_web3_log(
                     transaction_index: Some(receipt.tx_index.into()),
                     log_index:         Some((index + idx).into()),
                     removed:           false,
+                    log_type:          "".to_string(),
```
Comment (KaoImin) [inline core/api/src/jsonrpc/impl.rs]: Remove this.

File: core/api/src/jsonrpc/web3_types.rs [hunk 3]
```diff
@@ -444,6 +444,7 @@ pub struct Web3Filter {
     pub block_hash: Option<H256>,
     pub address:    Option<H160>,
     pub topics:     Option<Vec<H256>>,
+    pub limit:      Option<usize>,
```
Comment (KaoImin) [inline core/api/src/jsonrpc/web3_types.rs]: Remove this.

File: core/api/src/jsonrpc/web3_types.rs [hunk 4]
```diff
@@ -459,6 +460,8 @@ pub struct Web3Log {
     pub log_index:         Option<U256>,
     #[serde(default)]
     pub removed:           bool,
+    #[serde(rename = "type")]
```
Comment (KaoImin) [inline core/api/src/jsonrpc/web3_types.rs]: ditto

File: core/storage/src/lib.rs [hunk 5]
```diff
@@ -655,6 +655,19 @@ impl<Adapter: StorageAdapter> Storage for ImplStorage<Adapter> {
         let proof = ensure_get!(self, *LATEST_PROOF_KEY, LatestProofSchema);
         Ok(proof)
     }
+
+    async fn get_number_by_hash(
```
Comment (KaoImin) [inline core/storage/src/lib.rs]: ditto

File: protocol/src/traits/storage.rs [hunk 6]
```diff
@@ -118,6 +118,8 @@ pub trait Storage: CommonStorage {
     async fn update_latest_proof(&self, ctx: Context, proof: Proof) -> ProtocolResult<()>;
 
     async fn get_latest_proof(&self, ctx: Context) -> ProtocolResult<Proof>;
+
+    async fn get_number_by_hash(&self, ctx: Context, hash: &Hash) -> ProtocolResult<Option<u64>>;
```
Comment (KaoImin) [inline protocol/src/traits/storage.rs]: ditto

File: protocol/src/traits/api.rs [hunk 7]
```diff
@@ -82,4 +82,6 @@ pub trait APIAdapter: Send + Sync {
         position: U256,
         state_root: Hash,
     ) -> ProtocolResult<Bytes>;
+
+    async fn get_number_by_hash(&self, ctx: Context, hash: Hash) -> ProtocolResult<Option<u64>>;
```
Comment (KaoImin) [inline protocol/src/traits/api.rs]: ditto

File: devtools/chain/config.toml [hunk 8]
```diff
@@ -64,9 +64,9 @@ ckb_uri = "https://mercury-testnet.ckbapp.dev/rpc"
 mercury_uri = "http://127.0.0.1:8116"
 start_block_number = 4096800
 checkpoint_interval = 1000
-pk = "0x37aa0f893d05914a4def0460c0a984d3611546cfb26924d7a7ca6e0db9950a2d"
+pk = ""
```
Comment (KaoImin) [inline devtools/chain/config.toml]: revert this file

**Review by KaoImin — APPROVED** [review]

**Review by driftluo — APPROVED** [review]

**Pull request merged**